src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -782,7 +782,8 @@ public void run() {
                         }
                         listener.getLogger().println("Running on " + ModelHyperlinkNote.encodeTo(node) + " in " + workspace);
                         context.newBodyInvoker()
-                                .withContexts(exec, computer, env, workspace)
+                                .withContexts(exec, computer, env,
+                                    FilePathDynamicContext.createContextualObject(workspace))
```

Comment:
Since `DynamicContext.get` takes precedence over `StepContext.doGet`, what happens after this patch for something like:

```
node {
  dir('foo') {
    sh('pwd')
  }
}
```
Wouldn't the shell step use the `FilePath` from the node rather than the adjusted one from `dir` like it would today?

Reply:
No, but only because I also patched the `dir` step. See b0c0fa6a2a695a076377843cab8ddac8c765f25f for discussion.

Reply:
The `dir` and `ws` steps are distinct though, right? The following test fails for me with your changes, but if I revert src/main it passes:

```
diff --git a/src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java b/src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java
index 2c68d2d..78c9be4 100644
--- a/src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java
+++ b/src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java
@@ -1166,6 +1166,20 @@ public class ExecutorStepTest {
         });
     }
 
+    @Test public void dirSmokes() {
+        story.then(r -> {
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p1");
+            p.setDefinition(new CpsFlowDefinition(
+                    "node {\n" +
+                    "  dir('custom') {\n" +
+                    "    sh('echo \"$(pwd) is script working dir\"')\n" +
+                    "  }\n" +
+                    "}", true));
+            WorkflowRun b = r.buildAndAssertSuccess(p);
+            r.assertLogContains("/custom is script working dir", b);
+        });
+    }
+
```

Reply:
Ah sorry, yes that needs to be patched as well, in a separate PR. I will prepare it.

Reply:
I guess the bigger question is whether this matters for other types. I feel like `FilePath` is a bit special, and dynamic contexts for things like `TaskListenerDecorator` will avoid similar problems by being naturally additive with `TaskListenerDecorator.merge`, but it does seem a little concerning.

Reply:
> `FilePath` is a bit special

In that all uses of it as a contextual object would suffer from the same issue reported for `node` in JENKINS-41854.

> dynamic contexts for things like `TaskListenerDecorator` will avoid similar problems by being naturally additive with `TaskListenerDecorator.merge`

Not automatically. [Discussion](https://github.com/jenkinsci/workflow-support-plugin/blob/79b6e6a9b308da2cec0040a735826d3709df24c9/src/test/java/org/jenkinsci/plugins/workflow/support/DynamicContextTest.java#L109-L209)

> it does seem a little concerning

Agreed. I was hoping to keep this change relatively self-contained but I can also spend some time looking into whether explicit support in `workflow-cps` could improve the semantics.

Reply:
Is there any reason we can't just patch `DefaultStepContext` directly with something like this for `FilePath`? Will `get(Node.class)` fail if the `Computer` is no longer actually connected to the agent?

```
diff --git a/src/main/java/org/jenkinsci/plugins/workflow/support/DefaultStepContext.java b/src/main/java/org/jenkinsci/plugins/workflow/support/DefaultStepContext.java
index 4ccdbcc..243f1ee 100644
--- a/src/main/java/org/jenkinsci/plugins/workflow/support/DefaultStepContext.java
+++ b/src/main/java/org/jenkinsci/plugins/workflow/support/DefaultStepContext.java
@@ -25,6 +25,7 @@
 package org.jenkinsci.plugins.workflow.support;
 
 import hudson.EnvVars;
+import hudson.FilePath;
 import hudson.Launcher;
 import hudson.LauncherDecorator;
 import hudson.console.ConsoleLogFilter;
@@ -38,6 +39,7 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 import javax.annotation.CheckForNull;
 import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.FilePathUtils;
 import org.jenkinsci.plugins.workflow.flow.FlowExecution;
 import org.jenkinsci.plugins.workflow.flow.GraphListener;
 import org.jenkinsci.plugins.workflow.graph.FlowNode;
@@ -73,6 +75,16 @@ public abstract class DefaultStepContext extends StepContext {
             return key.cast(EnvironmentExpander.getEffectiveEnvironment(customEnvironment, (EnvVars) value, get(EnvironmentExpander.class), this, taskListener));
         } else if (key == Launcher.class) {
             return key.cast(makeLauncher((Launcher) value));
+        } else if (key == FilePath.class) {
+            FilePath fp = castOrNull(FilePath.class, value);
+            if (fp != null) {
+                Node n = get(Node.class);
+                if (n != null) {
+                    return castOrNull(key, FilePathUtils.find(n.getNodeName(), fp.getRemote()));
+                }
+            }
+            // Log warning?
+            return null;
         } else if (value != null) {
             return value;
         } else if (key == TaskListener.class) {
```

Reply:
The above assumes that the contextual `Node` is the owner of the contextual `FilePath`. Generally that will be true, but it is not something I would want to rely on.

I think what you meant is something more like

```java
// …as before
if (fp != null) {
    VirtualChannel vch = fp.getChannel();
    if (vch instanceof Channel && ((Channel) vch).isClosingOrClosed()) {
        String node = FilePathUtils.getNodeNameOrNull(fp);
        if (node != null) {
            FilePath fp2 = FilePathUtils.find(node, fp.getRemote());
            if (fp2 != null) {
                return fp2;
            }
        }
    }
}
```

That would likely suffice to solve JENKINS-41854, at the expense of being hackish. I would prefer to have a general solution to the issue of injecting contextual objects or dynamically refreshing them.

Another consideration is that replacing `FilePathPickle` by `FilePathDynamicContext.FilePathRepresentation` in `program.dat` is a step away from relying on pickles, and thus the close coupling between restart semantics and Groovy engine internals. For example, a program cannot resume running until its `FilePathPickle` has been resolved, so in a program like

```groovy
parallel linux: {
  node('linux') {
    timeout(60) {
      sh 'make world'
    }
  }
}, win: {
  node('win') {
    timeout(60) {
      bat 'make world'
    }
  }
}
```

if Jenkins is restarted in the middle of both builds, and the Windows agent fails to reconnect, the build will hang indefinitely and you will not even be able to see if the Linux build has completed. Using `DynamicContext` only for `FilePath` will not solve this (there are other pickles which will hang here too), but it is a start.

Reply:
> Another consideration is that replacing `FilePathPickle` by `FilePathDynamicContext.FilePathRepresentation` in `program.dat`...

Interesting, I missed that this means `FilePathPickle` wouldn't be used for new builds (unless some other plugin is adding a FilePath directly as a context variable). For builds across the restart, everything should be ok because `FilePathDynamicContext` won't find a `FilePathRepresentation` and will return null, and so `StepContext.doGet` will be used, right?

Reply:
> For builds across the restart

You mean, when updating plugins. Yes that is how it should work—`program.dat` will still contain a `FilePathPickle`, `FilePathDynamicContext.get` will return `null`, and `DefaultStepContext.get` will go back to checking `CpsStepContext.doGet(FilePath.class)`. Of course in this case JENKINS-41854 will still be observable:

```groovy
node {
  sh 'sleep 1h' // restarting during this
  sleep 9999 // disconnect agent during this
  sh ':' // would fail due to invalid Channel loaded during FilePathPickle.rehydrate
}
```